File: java/easy/majorityElement.java
// Took the Boyer-Moore Voting Algorithm approach. You set the value you're looping through as the candidate.
// If the next value in the loop is the candidate increment the count of the item.
// If the next value isn't decrement the count.
// If count is 0 then set the new candidate as that value.
// At the end of the loop the majority value should be found.

class Solution {
  public int majorityElement(int[] nums) {

      int candidate = 0;
      int count = 0;

      for (int i = 0; i < nums.length; i++) {

          if (count == 0) {
              candidate = nums[i];
              count++;
              continue;
          };

          if (candidate == nums[i]) {
              count++;
              continue;
          } else {
              count--;
              continue;
          }
      };

      return candidate;
  }
}

File: python/easy/majority_element.py
puzzle = 'https://leetcode.com/problems/majority-element/description/'

# brute force solution and first attempt, it does pass the test but runs slower than ideal:
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        unique_elements = {}

        for item in nums:
            if item in unique_elements:
                unique_elements[item] += 1
            else:
                unique_elements[item] = 0

        highest_key = 0
        highest_value = 0
        
        for key, value in unique_elements.items():
            if value > highest_value:
                highest_key = key
                highest_value = value
            else:
                continue

        return highest_key
    
# this solution isn't super efficient as it has a space and time complexity of O(n),
# which means it will get more complex as the list grows.
# Running two loops is not an ideal solution for this problem.

# optimal Boyer-Moore Voting algorithm for this problem, this is the more efficient solution,
# however it ran just as fast as the brute force solution on leetcode:

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        candidate = None
        count = 0

        for num in nums:
            if count == 0:
                candidate = num
                count = 1
            elif num == candidate:
                count += 1
            else:
                count -= 1

        return candidate
    
